### PR TITLE
feat(maven-wrapper): support wrapperVersion

### PR DESCRIPTION
**Pull request by trystian1**
… standard after 3.3.0

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/maven-wrapper/extract.spec.ts [hunk 1]
```diff
@@ -2,11 +2,19 @@ import { extractPackageFile } from '.';
 
 const onlyWrapperProperties =
   'wrapperUrl=https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/0.5.6/maven-wrapper-0.5.6.jar';
+
+const onlyWrapperVersionProperties =
+  'wrapperVersion=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/wrapper/maven-wrapper/3.3.0/maven-wrapper-3.3.0.jar';
```
Comment (breun) [inline lib/modules/manager/maven-wrapper/extract.spec.ts:7]: The value of the `wrapperVersion` property should be just a version number, so for example: `wrapperVersion=3.3.0`.

File: lib/modules/manager/maven-wrapper/extract.ts [hunk 2]
```diff
@@ -15,10 +15,17 @@ const WRAPPER_URL_REGEX = regEx(
   '^(?:wrapperUrl\\s*=\\s*)(?<url>\\S*-(?<version>\\d+\\.\\d+(?:\\.\\d+)?(?:-\\w+)*)(?:.jar))',
 );
 
+const WRAPPER_VERSION_REGEX = regEx(
+  '^(?:wrapperVersion\\s*=\\s*)(?<url>\\S*-(?<version>\\d+\\.\\d+(?:\\.\\d+)?(?:-\\w+)*)(?:.jar))',
```
Comment (breun) [inline lib/modules/manager/maven-wrapper/extract.ts:19]: The value of the `wrapperVersion` property should be just a version number, so for example: `wrapperVersion=3.3.0`. This shouldn't need any regular expression or other parsing.

File: lib/modules/manager/maven-wrapper/extract.spec.ts (hunk 1 above)
Comment (breun) [reply lib/modules/manager/maven-wrapper/extract.spec.ts:7]: The trickiest case is `maven-wrapper-3.3.0-only-script`, because the has `wrapperUrl` nor `wrapperVersion`:

```
distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.9.6/apache-maven-3.9.6-bin.zip
```

I believe the Maven Wrapper version for `maven-wrapper-3.3.0-only-script` can only be detected by looking for the string `Apache Maven Wrapper startup batch script, version 3.3.0` in `mvnw` (or `mvnw.cmd` if `mvnw` is not present).

Comment (breun) [reply lib/modules/manager/maven-wrapper/extract.spec.ts:7]: I'm currently trying to find out if Maven Wrapper provides a way to update Maven Wrapper using the existing installation type, without having to explicitly specify (and thus determine) it. If so, Renovate should update Maven Wrapper like that.

If upgrading using the existing installation type can only be done by specifying the type in the update command (`-Dtype=${type}`), then I plan to request a property for that too (e.g. `wrapperType`). But for older versions that don't explicitly communicate their installation type, Renovate might still need to detect the installation type. I suggest keeping this installation type business out of scope for this current PR. That could be added later if needed.

Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.spec.ts:7]: How does it figure out on which url it should get the maven-wrapper ? Or is that now all covered in mvn ?

File: lib/modules/manager/maven-wrapper/extract.spec.ts [hunk 3]
```diff
@@ -2,11 +2,19 @@ import { extractPackageFile } from '.';
...
+
 const onlyMavenProperties =
   'distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip';
 
 const wrapperAndMavenProperties = `distributionUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip\nwrapperUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar`;
 
+const wrapperVersionAndMavenProperties = `distributionUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip\nwrapperVersion=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/wrapper/maven-wrapper/3.3.0/maven-wrapper-3.3.0.jar`;
```
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.spec.ts:14]: please use `codeBlock` template helper for better readability. use only for your new tests

File: lib/modules/manager/maven-wrapper/extract.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/manager/maven-wrapper/extract.ts:19]: it seems we only need the version group? 🤔

**Comment by rarkins** [comment]
@breun thanks again for this PR. What can we do to help get it mergeable?

**Comment by breun** [comment]
> @breun thanks again for this PR. What can we do to help get it mergeable?

I’ve reached out to @trystian1, and I do believe he is still willing to help, but he hasn’t yet been able to find time to work on this further.

I also believe he wouldn’t mind if someone else implemented the changes required for the desired behavior I described.

File: lib/modules/manager/maven-wrapper/extract.ts (hunk 2 above)
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.ts]: yeah so I pushed an update but that will leave the url empty,
which I think is no problem because `wrapper:wrapper` will do the work.

But I'm not sure how we are going to make this work (and if we should). I'm in no way a maven-wrapper expert so I will discuss with Nils if this a problem.

```
function getCustomMavenWrapperRepoUrl(
  deps: PackageDependency[],
): string | null {
  const replaceString = deps.find(
    (dep) => dep.depName === 'maven-wrapper',
  )?.replaceString;

  if (!replaceString) {
    return null;
  }

  const match = regEx(/^(.*?)\/org\/apache\/maven\/wrapper\//).exec(
    replaceString,
  );

  if (!match) {
    return null;
  }

  return match[1] === DEFAULT_MAVEN_REPO_URL ? null : match[1];
}
```

Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.ts]: since it will give that as option to the maven command:

```
function getExtraEnvOptions(deps: PackageDependency[]): ExtraEnv {
  const customMavenWrapperUrl = getCustomMavenWrapperRepoUrl(deps);
  if (customMavenWrapperUrl) {
    return { MVNW_REPOURL: customMavenWrapperUrl };
  }
  return {};
}
```

File: lib/modules/manager/maven-wrapper/extract.spec.ts [hunk 4]
```diff
@@ -2,11 +2,18 @@ import { extractPackageFile } from '.';
...
   'wrapperUrl=https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/0.5.6/maven-wrapper-0.5.6.jar';
+
+const onlyWrapperVersionProperties = 'wrapperVersion=3.3.8';
+
 const onlyMavenProperties =
   'distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip';
 
 const wrapperAndMavenProperties = `distributionUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip\nwrapperUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar`;
 
+const wrapperVersionAndMavenProperties = `distributionUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip\nwrapperVersion=3.3.8`;
```
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.spec.ts:13]:
```suggestion
const wrapperVersionAndMavenProperties = codeBlock`
  distributionUrl=https://internal.artifactory.acme.org/artifactory/maven-bol/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip
  wrapperVersion=3.3.8
  `;
```

please use codeblock template helper for new fixtures. inline them to the test if only used once

File: lib/modules/manager/maven-wrapper/extract.ts [hunk 5]
```diff
@@ -15,10 +15,16 @@ const WRAPPER_URL_REGEX = regEx(
   '^(?:wrapperUrl\\s*=\\s*)(?<url>\\S*-(?<version>\\d+\\.\\d+(?:\\.\\d+)?(?:-\\w+)*)(?:.jar))',
 );
 
+const WRAPPER_VERSION_REGEX = regEx(/wrapperVersion=(?<version>\d+(\.\d+)*)/);
```
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:18]: are you sure those versions are allowed:
- `3`
- `3.8`
- `3.8.1.2`
i think this is better:
```suggestion
const WRAPPER_VERSION_REGEX = regEx(/wrapperVersion=(?<version>\\d+\\.\\d+(?:\\.\\d+)?(?:-\\w+)*)/);
```
use the same regex like on the url and support prereleases

File: lib/modules/manager/maven-wrapper/extract.ts [hunk 6]
```diff
@@ -28,8 +34,8 @@ function extractLineInfo(lines: string[], regex: RegExp): Version | null {
       const match = regex.exec(line);
       if (match?.groups) {
         return {
-          url: match.groups.url,
-          version: match.groups.version,
+          url: match.groups.url ?? null,
+          version: match.groups.version ?? null,
```
Comment (viceice) [inline lib/modules/manager/maven-wrapper/extract.ts:38]: why this change? looks like it's not required

File: lib/modules/manager/maven-wrapper/extract.ts (hunk 5 above)
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.ts:18]: I tought my regex did support it, but fine with yours as wel!

File: lib/modules/manager/maven-wrapper/extract.ts (hunk 6 above)
Comment (trystian1) [reply lib/modules/manager/maven-wrapper/extract.ts:38]: can revert that indeed, 
I tought it was nicer that url if it was null instead of undefined when not there

Comment (viceice) [reply lib/modules/manager/maven-wrapper/extract.ts:38]: please revert changes.

**Comment by rarkins** [comment]
@trystian1 do you have time available to resume this PR?

**Comment by nhmarujo** [comment]
@rarkins I just saw you closed this one. Does it mean this is not getting fixed? Or was it fixed elsewhere? Thanks

**Comment by rarkins** [comment]
It's closed due to inactivity and is not implemented. You're welcome to submit your own PR instead

**Comment by nhmarujo** [comment]
Thanks @rarkins . Unfortunately I don't know much about the rules here. @trystian1 looked to be on a good path, sad it didn't move forward.
